luaui/Widgets/gui_attack_aoe.lua

```diff
@@ -79,127 +65,413 @@ local glDeleteList = gl.DeleteList
+		local attackNoEnemy = WG['attacknoenety']
+
+		-- Determine if we should ignore this unit based on mode (dgun vs attack)
+		local ignoreAlly = (dgun and dgunNoAlly) or (not dgun and attackNoAlly)
+		local ignoreEnemy = (dgun and dgunNoEnemy) or (not dgun and attackNoEnemy)
+
+		if isAlly and ignoreAlly then
+			forceGround = true
+		elseif not isAlly and ignoreEnemy then
+			local unitDefID = spGetUnitDefID(target)
+			local uc = Cache.UnitProperties
+			local isPassThrough = uc.isAir[unitDefID] or uc.isShip[unitDefID] or uc.isUnderwater[unitDefID]
+
+			if not isPassThrough and uc.isHover[unitDefID] then
+				local _, pos = spTraceScreenRay(tx, ty, true)
+				if pos and spGetGroundHeight(pos[1], pos[3]) < 0 then
+					isPassThrough = true
 				end
-			else
-				return GetUnitPosition(mouseTarget)
 			end
-		elseif mouseTargetType == "feature" then
-			local mouseTargetType, mouseTarget = TraceScreenRay(mx, my, true)
-			if mouseTarget then
-				return mouseTarget[1], mouseTarget[2], mouseTarget[3]
+
+			if not isPassThrough then
+				forceGround = true
+			else
+				return spGetUnitPosition(target)
 			end
 		else
-			return nil
+			return spGetUnitPosition(target)
+		end
+	end
+
+	if forceGround then
+		local _, groundTarget = spTraceScreenRay(tx, ty, true)
+		if groundTarget then
+			return groundTarget[1], groundTarget[2], groundTarget[3]
 		end
-	else
 		return nil
 	end
+
+	return target[1], target[2], target[3]
 end
 
 local function GetMouseDistance()
-	local cx, cy, cz = GetCameraPosition()
-	local mx, my, mz = GetMouseTargetPosition()
-	if not mx then
+	local cx, cy, cz = spGetCameraPosition()
+	local tx, ty, tz = GetMouseTargetPosition()
+	if not tx then
 		return nil
 	end
-	local dx = cx - mx
-	local dy = cy - my
-	local dz = cz - mz
-	return sqrt(dx * dx + dy * dy + dz * dz)
+	return distance3d(cx, cy, cz, tx, ty, tz)
+end
+
+--------------------------------------------------------------------------------
+-- UTILITY FUNCTIONS
+--------------------------------------------------------------------------------
+local function FadeColorInPlace(color, alphaMult)
+	color[4] = color[4] * alphaMult
+end
+
+local function SetGlColor(alphaFactor, color)
+	glColor(color[1], color[2], color[3], color[4] * alphaFactor)
+end
+
+local function LerpColor(sourceColor, targetColor, t, out)
+	out[1] = lerp(sourceColor[1], targetColor[1], t)
+	out[2] = lerp(sourceColor[2], targetColor[2], t)
+	out[3] = lerp(sourceColor[3], targetColor[3], t)
+	out[4] = lerp(sourceColor[4], targetColor[4], t)
+	return out
+end
+
+local function CopyColor(source, target)
+	target[1] = source[1]
+	target[2] = source[2]
+	target[3] = source[3]
+	target[4] = source[4]
+end
+
+local function ToBool(x)
+	return x ~= nil and x and x ~= 0 and x ~= "false"
+end
+
+local function GetNormalizedAndMagnitude(x, y, z)
+	local mag = distance3d(x, y, z, 0, 0, 0)
+	if mag == 0 then
+		return nil
+	else
+		return x / mag, y / mag, z / mag, mag
+	end
```

Comment:
In the same spirit as above:

```suggestion
	if mag ~= 0 then
		return x / mag, y / mag, z / mag, mag
	end
```

Is lighter and faster, just barely.